components/authz-service/engine/opa/opa_internal_test.go

```diff
@@ -549,3 +429,864 @@ func randomPolicies(i int) []map[string]interface{} {
+			var resp []string
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2ProjectsAuthorized(ctx, []string{"user:local:test"}, "compliance:profiles:list", "compliance:profiles", allProjects)
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			projectsResponse = resp
+		})
+	}
+}
+
+func BenchmarkV2FilterAuthorizedProjectsWithIncreasingPolicies(b *testing.B) {
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	policyCount := []int{0, 5, 10, 20, 50, 100, 200, 1000}
+	roleCount := 10 // keep this constant while increasing policyCount
+
+	for _, count := range policyCount {
+		policies, roles := v2BaselineAndRandomPoliciesAndRoles(count, roleCount)
+
+		s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+			"policies": policies,
+			"roles":    roles,
+		})
+		err = s.makeAuthorizedProjectPreparedQuery(ctx)
+		require.NoError(b, err, "update OPA store and prepare projects query")
+
+		b.Run(fmt.Sprintf("store with %d custom policies and %d custom roles", count, roleCount), func(b *testing.B) {
+			var resp []string
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2FilterAuthorizedProjects(ctx, []string{"user:local:test"})
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			projectsResponse = resp
+		})
+	}
+}
+
+func BenchmarkV2AuthorizedProjectPreparedQueryWithIncreasingRoles(b *testing.B) {
+	var r error
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	policyCount := 20 // keep this constant while increasing roleCount
+	roleCount := []int{0, 5, 10, 20, 50, 100}
+
+	chefPolicies, _ := v2BaselinePoliciesAndRoles()
+
+	for _, count := range roleCount {
+		policies, roles := v2BaselineAndRandomPoliciesAndRoles(policyCount, count)
+		s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+			"policies": policies,
+			"roles":    roles,
+		})
+
+		b.Run(fmt.Sprintf("store with %d chef-managed policies and %d custom policies", len(chefPolicies), count),
+			func(b *testing.B) {
+				for n := 0; n < b.N; n++ {
+					r = s.makeAuthorizedProjectPreparedQuery(ctx)
+					if r != nil {
+						b.Error(r)
+					}
+				}
+				errResult = r
+			})
+	}
+}
+
+func BenchmarkV2ProjectsAuthorizedWithIncreasingRoles(b *testing.B) {
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	policyCount := 20 // keep this constant while increasing roleCount
+	roleCount := []int{0, 5, 10, 20, 50, 100}
+
+	for _, count := range roleCount {
+		policies, roles := v2BaselineAndRandomPoliciesAndRoles(policyCount, count)
+		s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
```

Comment:
the `v2p1Store` contains all the policy and role information that OPA needs to answer authorization queries.